core/src/avm2/optimize.rs

```diff
@@ -0,0 +1,893 @@
+            | Op::Kill { index }
+            | Op::IncLocal { index }
+            | Op::IncLocalI { index }
+            | Op::DecLocal { index }
+            | Op::DecLocalI { index } => {
+                if (*index as usize) < initial_local_types.len() {
+                    initial_local_types.set_any(*index as usize);
+                }
+            }
+            Op::HasNext2 {
+                object_register,
+                index_register,
+            } => {
+                if (*object_register as usize) < initial_local_types.len() {
+                    initial_local_types.set_any(*object_register as usize);
+                }
+                if (*index_register as usize) < initial_local_types.len() {
+                    initial_local_types.set_any(*index_register as usize);
+                }
+            }
+            _ => {}
+        }
+    }
+
+    let mut stack = Stack::new();
+
+    macro_rules! stack_pop_multiname {
+        ($index: expr) => {{
+            let multiname = method
+                .translation_unit()
+                // note: ideally this should be a VerifyError here or earlier
+                .pool_maybe_uninitialized_multiname(*$index, &mut activation.context);
+
+            if let Ok(multiname) = multiname {
+                stack.pop_for_multiname(multiname);
+                Some(multiname)
+            } else {
+                None
+            }
+        }};
+    }
+
+    let mut scope_stack = Stack::new();
+    let mut local_types = initial_local_types.clone();
+
+    for (i, op) in code.iter_mut().enumerate() {
+        if jump_targets.contains(&(i as i32)) {
+            stack.clear();
+            scope_stack.clear();
+            local_types = initial_local_types.clone();
+        }
+
+        match op {
+            Op::CoerceB => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Boolean)) {
+                    *op = Op::Nop;
+                }
+                stack.push_boolean();
+            }
+            Op::CoerceD => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Number)) {
+                    *op = Op::Nop;
+                }
+                stack.push_number();
+            }
+            Op::CoerceI => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Int))
+                    || matches!(stack_value, Some(ValueType::Uint))
+                {
+                    *op = Op::Nop;
+                }
+                stack.push_int();
+            }
+            Op::CoerceU => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Uint)) {
+                    *op = Op::Nop;
+                }
+                stack.push_uint();
+            }
+            Op::CoerceA => {
+                stack.pop();
+                stack.push_any();
```

Comment:
Can we document that this appears to inhibit optimizations in FP? In case someone decides to change this blindly in the future.